mycroft/enclosure/hardware_enclosure.py

```diff
@@ -0,0 +1,137 @@
+# Copyright 2020 Mycroft AI Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import importlib
+import threading
+import time
+from mycroft.util import create_signal
+
+class HardwareEnclosure:
+    mute_led = 11    # last led is reserved for mute mic switch
+
+    def __init__(self, enclosure_type, board_type):
+        self.enclosure_type = enclosure_type
+        self.board_type = board_type
+
+        self.max_volume = 10
+        self.shadow_volume = 5
+        self.volume_increment = 1
+        self.last_action = time.time()
+        self.last_mute = 0
+
+        driver_dir = "mycroft.enclosure.hardware"
+
+        capabilities_module = driver_dir + ".%s" % (self.enclosure_type,)
+        module = importlib.import_module(capabilities_module)
+        enclosure_capabilities = module.Capabilities()
+        if self.board_type is None:
+            self.board_type = enclosure_capabilities.board_type
+        self.capabilities = enclosure_capabilities.capabilities[self.board_type]
+
+        # eventually this could also be driven by the capabilities
+        led_driver = self.capabilities["Led"]["name"]
+        switch_driver = self.capabilities["Switch"]["name"]
+        volume_driver = self.capabilities["Volume"]["name"]
+
+        pal_module = driver_dir + ".MycroftLed.%s" % (self.capabilities["Palette"]["name"],)
+        module = importlib.import_module(pal_module)
+        self.palette = module.Palette()
+
+        led_module = driver_dir + ".MycroftLed.%s" % (led_driver,)
+        module = importlib.import_module(led_module)
+        self.leds = module.Led()
+
+        vol_module = driver_dir + ".MycroftVolume.%s" % (volume_driver,)
+        module = importlib.import_module(vol_module)
+        self.hardware_volume = module.Volume()
+
+        switch_module = driver_dir + ".MycroftSwitch.%s" % (switch_driver,)
+        module = importlib.import_module(switch_module)
+        self.switches = module.Switch()
+
+        self.switches.user_action_handler = self.handle_action
+        self.switches.user_mute_handler = self.handle_mute
+        self.switches.user_volup_handler = self.handle_vol_up
+        self.switches.user_voldown_handler = self.handle_vol_down
+
+        # TODO - pull up/down verified!!!
+        self.leds._set_led(self.mute_led, self.palette.GREEN)
+
+        # volume display timeout
+        self.watchdog = None
+        self.watchdog_timeout = 5
+
+    def get_capabilities(self):
+        return self.capabilities
```

Comment:
For the broader range of hardware enclosures, should this include other capabilities like:
- GUI vs voice only
- Mouth and/or Eyes (eg Mark I)
- WiFi
- System actions - reboot, shutdown, update, ssh
- other?

Potentially the capabilities should have a base class with everything defaulted to None?

Reply:
This code handles the three unique features of the sj201; leds, switches and volume control. If I had more time I would greatly enhance the base.py enclosure code but there are only so many hours in a day. Since you are recommending constants which will not be used, feel free to have at it.